server/src/main/java/org/elasticsearch/index/mapper/BlockLoader.java

```diff
@@ -0,0 +1,191 @@
+     * It's important for performance that this only have one implementation in
+     * production code. That implementation sits in the "compute" project. The is
+     * also a test implementation, but there may be no more other implementations.
+     */
+    interface BuilderFactory {
+        /**
+         * Build a builder to load booleans as loaded from doc values. Doc values
+         * load booleans deduplicated and in sorted order.
+         */
+        BooleanBuilder booleansFromDocValues(int expectedCount);
+
+        /**
+         * Build a builder to load booleans without any loading constraints.
+         */
+        BooleanBuilder booleans(int expectedCount);
+
+        /**
+         * Build a builder to load {@link BytesRef}s as loaded from doc values.
+         * Doc values load {@linkplain BytesRef}s deduplicated and in sorted order.
+         */
+        BytesRefBuilder bytesRefsFromDocValues(int expectedCount);
+
+        /**
+         * Build a builder to load {@link BytesRef}s without any loading constraints.
+         */
+        BytesRefBuilder bytesRefs(int expectedCount);
+
+        /**
+         * Build a builder to load doubles as loaded from doc values.
+         * Doc values load doubles deduplicated and in sorted order.
+         */
+        DoubleBuilder doublesFromDocValues(int expectedCount);
+
+        /**
+         * Build a builder to load doubles without any loading constraints.
+         */
+        DoubleBuilder doubles(int expectedCount);
+
+        /**
+         * Build a builder to load ints as loaded from doc values.
+         * Doc values load ints deduplicated and in sorted order.
+         */
+        IntBuilder intsFromDocValues(int expectedCount);
+
+        /**
+         * Build a builder to load ints without any loading constraints.
+         */
+        IntBuilder ints(int expectedCount);
+
+        /**
+         * Build a builder to load longs as loaded from doc values.
+         * Doc values load longs deduplicated and in sorted order.
+         */
+        LongBuilder longsFromDocValues(int expectedCount);
+
+        /**
+         * Build a builder to load longs without any loading constraints.
+         */
+        LongBuilder longs(int expectedCount);
+
+        /**
+         * Build a builder that can only load null values.
+         * TODO this should return a block directly instead of a builder
+         */
+        Builder nulls(int expectedCount);
+
+        /**
+         * Build a reader for reading keyword ordinals.
+         */
+        SingletonOrdinalsBuilder singletonOrdinalsBuilder(SortedDocValues ordinals, int count);
+
+        // TODO support non-singleton ords
+    }
+
+    /**
+     * A builder for typed values. For each document you may either call
+     * {@link #appendNull}, {@code append<Type>}, or
+     * {@link #beginPositionEntry} followed by two or more {@code append<Type>}
+     * calls, and then {@link #endPositionEntry}.
+     */
+    interface Builder {
```

Comment:
I think we should make this Builder Releasable and release it if we hit a breaker when reading values. Let's do this in a follow-up.

Reply:
++. I think we *have* to do it to properly track memory on blocks built by field loading. We just aren't doing that yet.